### PR TITLE
[Issue #432] Limit number of results returned

### PR DESCRIPTION
**Pull request by haochen07**
This PR related to Issue #432 

**Note**
- Updated json parser to parse limit number.
- Checked the number of results each time before returning them. Once reaching the limit, send `Done` message and stop querying.

**Test**
- Added test case to Json parser and cloudberry client.
- All unit test cases pass.
- Tested on current TwitterMap.

### CODE REVIEW

**Comment by haochen07** [comment]
Sorry, I don't have authority to add reviewers. @JavierJia @waans11 @baiqiushi You can take a look at this ongoing PR.

**Comment by JavierJia** [comment]
LGTM. I will let @waans11 and @baiqiushi to review the details.

File: cloudberry/zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/JSONParser.scala [hunk 1]
```diff
@@ -56,6 +56,41 @@ object JSONParser {
     resolved
   }
 
+  /**
+    * Parses json request and extract [[QueryExeOption]], then removes all option-related field from requests.
+    *
+    * @param json
+    * @return
+    */
+  def extractQueryExeOption(json: JsValue): (JsValue, QueryExeOption) = {
```
Comment (JavierJia) [inline cloudberry/zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/JSONParser.scala]: the logic of this function seems hard to follow. Could you elaborate the logic in the ScalaDoc? 
And if the logic is more appropriate in the `if else` block, we'd better write the logic in that way than the `case` way.

File: cloudberry/zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/schema/Query.scala [hunk 2]
```diff
@@ -15,15 +15,23 @@ trait IWriteQuery extends IQuery {
 
 }
 
-case class QueryExeOption(sliceMills: Int,
-                          continueSeconds: Int)
+case class JsonRequestOption(sliceMills: Int,
```
Comment (JavierJia) [inline cloudberry/zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/schema/Query.scala]: I think this is still the `QueryExeOption`. It is the parsed object result, like `Query`, not the `JSON`, right?

File: cloudberry/zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/schema/Query.scala [hunk 3]
```diff
@@ -15,15 +15,23 @@ trait IWriteQuery extends IQuery {
...
 
-object QueryExeOption {
-  val NoSliceNoContinue = QueryExeOption(-1, -1)
+object JsonRequestOption {
+  val NoSliceNoContinue = JsonRequestOption(-1, -1)
   val TagSliceMillis = "sliceMillis"
   val TagContinueSeconds = "continueSeconds"
 }
 
+case class QueryExeOption(sliceMills: Int,
```
Comment (JavierJia) [inline cloudberry/zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/schema/Query.scala]: Now I confused, why do we need this two almost identical options?

File: cloudberry/zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/JSONParser.scala (hunk 1 above)
Comment (haochen07) [reply cloudberry/zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/JSONParser.scala]: The parsing logic is
- if there is no `option`, treat queries by the original way (meaning if there is `limit` in the query, it will be executed in database statement not controlled by cloudberry, just as it is today)

- if there is `option`,

  - for `batch` queries, if `limit` presents, throw errors. Because now twittermap use `limit` in `batch` (the point we were missing in the previous discussion), there is one specialization for this: we enable `aggregate` and `limit` present together.

  - for single query, if `limit` presents, this is the case we are targeting. We allow two possible APIs.
```
{
    "select" : {
        "limit": 100
    }
}
```
and 
```
{
    "select" : {
        "order" : [ "-create_at"],  // required
        "limit": 100,                       // required
        "offset" : 0,                       // required
        "field": ["create_at", "id"] // optional
    }
}
```
for the first one, we remove `select` field after parsing the `limit` value to `QueryExeOption` and for the second one we take the limit value and reassign `Int.MaxValue` to `limit`. This way we keep the format constrains in `select` statement and parse the `limit` as well.

File: cloudberry/zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/schema/Query.scala (hunk 3 above)
Comment (haochen07) [reply cloudberry/zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/schema/Query.scala]: Because the `limit` is in `select` field somewhere in the query, and all other options are in the `option` field. We need a case class to parse the `option` field and we also need an object that contains all options including `limit`. That is the intuitive.

**Comment by haochen07** [comment]
Remove the parsing part to next PR. This one only contains the control logic. Temporarily put the `limit` API into `option`.

**Comment by haochen07** [comment]
@waans11 and @baiqiushi , do you have any comments or something looks unfamiliar and need me to explain for this PR?

**Comment by JavierJia** [comment]
@haochen07 Let's merge this PR and move on to the second part.
